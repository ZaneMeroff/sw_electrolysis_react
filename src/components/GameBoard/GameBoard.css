.gameboard-outer-container {
  align-items: center;
  display: flex;
  height: 340px;
  justify-content: space-around;
  margin: 1%;
  margin-top: 30px;
  width: 340px;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
	-tap-highlight-color: rgba(0,0,0,0);
}

.center-circle {
  background-color: #313131;
  border-radius: 70px;
  display: flex;
  flex-direction: column;
  height: 110px;
  justify-content: center;
  position: absolute;
  width: 110px;
  z-index: 3;
}

.turn-text-1 {
  color: #FFF;
  margin-bottom: 0px;
  margin-top: 4px;
  font-family: 'Press Start 2P', cursive;
  font-size: 90%;
}

.turn-text-2 {
  color: #FFF;
  font-family: 'Press Start 2P', cursive;
  font-size: 90%;
  margin-top: 10px;
}

.red-box,
.green-box,
.blue-box,
.yellow-box {
  height: 170px;
  width: 170px;
}

.red-box:focus:not(:focus-visible),
.green-box:focus:not(:focus-visible),
.blue-box:focus:not(:focus-visible),
.yellow-box:focus:not(:focus-visible) {
  outline: none;
}

.red-box {
  background-color: red;
  border: 3px solid #313131;
  border-top-left-radius: 200px;
  opacity: .3;
}

.red-box:hover {
  cursor: pointer;
}

.red-active,
.red-box:active {
  border: 3px solid red;
  box-shadow: 0 0 40px 20px red;
  opacity: 1;
}

.green-box {
  background-color: #00ff00;
  border: 3px solid #313131;
  border-top-right-radius: 200px;
  opacity: .3;
}

.green-box:hover {
  cursor: pointer;
}

.green-active,
.green-box:active {
  border: 3px solid #00ff00;
  box-shadow: 0 0 40px 20px #00ff00;
  opacity: 1;
}

.blue-box {
  background-color: #006bff;
  border: 3px solid #313131;
  border-bottom-right-radius: 200px;
  opacity: .3;
}

.blue-box:hover {
  cursor: pointer;
}

.blue-active,
.blue-box:active {
  border: 3px solid #006bff;
  box-shadow: 0 0 40px 20px #006bff;
  opacity: 1;
}

.yellow-box {
  background-color: yellow;
  border: 3px solid #313131;
  border-bottom-left-radius: 200px;
  opacity: .3;
}

.yellow-box:hover {
  cursor: pointer;
}

.yellow-active,
.yellow-box:active {
  border: 3px solid yellow;
  box-shadow: 0 0 40px 20px yellow;
  opacity: 1;
}

.disabled {
  pointer-events: none;
}

.top-section,
.bottom-section {
  display: flex;
  flex-direction: column;
}

.rules-outer-container {
  align-items: center;
  border: 2px dotted #FFF;
  display: flex;
  height: 340px;
  justify-content: center;
  margin-top: 10px;
  width: 340px;
}

.rules-inner-container {
  height: 90%;
  width: 90%;
}

.rules-text {
  color: #FFF;
  font-family: 'Press Start 2P', cursive;
  font-size: 90%;
  line-height: 25px;
}
