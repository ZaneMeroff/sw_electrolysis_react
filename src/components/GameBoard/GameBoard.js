import React, { useState } from 'react';
import { ScoreBoard } from '../ScoreBoard/ScoreBoard';
import './GameBoard.css';

let randomNumArray = [];
let playerResponse = [];
let playerClickCounter = 0;
let roundCount = 0;
let j;

export const GameBoard = () => {

  const [redActive, setRedActive] = useState(false);
  const [yellowActive, setYellowActive] = useState(false);
  const [greenActive, setGreenActive] = useState(false);
  const [blueActive, setBlueActive] = useState(false);
  const [myOrYours, setMyOrYours] = useState('');
  const [turnText, setTurnText] = useState('');
  const [playerClickDisplay, setPlayerClickDisplay] = useState(0);
  const [disableButtons, setDisableButtons] = useState(true);
  const [rulesVisible, setRulesVisible] = useState(false);
  const [gameBoardVisible, setGameBoardVisible] = useState(true);
  const [disableScoreBoard, setDisableScoreBoard] = useState(false);   

  const colorKey = {
    0 : setRedActive,
    1 : setYellowActive,
    2 : setGreenActive,
    3 : setBlueActive
  }

  const startGame = () => {
    randomNumArray = generate10randomNumbers();
    roundCount = 0;
    playerResponse = [];
    playerClickCounter = 0;
    setPlayerClickDisplay(0);
    setMyOrYours('MY');
    setTurnText('TURN!');
    nextRound();
    setDisableScoreBoard(true);     
  }

  const onViewRulesClick = () => {
    if (!rulesVisible) {
      setRulesVisible(true);
      setGameBoardVisible(false);
    } else {
      setRulesVisible(false);
      setGameBoardVisible(true);
    }
  }

  const validateAnswer = () => {
    let correctSoFar = true;
    for (let i = 0; i < playerResponse.length; i++) {
      if (playerResponse[i] !== randomNumArray[i] && correctSoFar) {
        correctSoFar = false;
      }
    }
    if (correctSoFar) {
      onCorrectRoundGuess();
    } else {
      gameOver();
    }
  }

  const gameOver = () => {
    setMyOrYours('GAME');
    setTurnText('OVER!');
    setDisableButtons(true);
    setDisableScoreBoard(false);  
  }

  const onCorrectRoundGuess = () => {
    setMyOrYours('RIGHT!');
    setTurnText('');
    setDisableButtons(true);
    setDisableScoreBoard(true);    
    playerResponse = [];
    setTimeout(() => {
      nextRound();
    }, 1000)
  }

  const nextRound = () => {
    if (roundCount === 10) {
      displayWinningMessage();
    } else {
      roundCount++
      playerClickCounter = 0
      setPlayerClickDisplay(0);
      setMyOrYours('MY');
      setTurnText('TURN!');
      j = 0;
      for (let i = 1; i <= (roundCount * 2); i++) {
        makeColorFlash(i);
      }
    }
  }

  const displayWinningMessage = () => {
    setMyOrYours('YOU');
    setTurnText('WIN!');
    setDisableButtons(true);
  }

  const makeColorFlash = i => {
    setTimeout(() => {
      if (i % 2) {
        colorKey[randomNumArray[Math.floor(j)]](true)
      } else {
        colorKey[randomNumArray[Math.floor(j)]](false)
      }
      j += 0.5
      if (i === (roundCount * 2)) {
        setMyOrYours('YOUR');
        setDisableButtons(false);
        setDisableScoreBoard(false);    
      }
    }, i * 800 );
  }

  const generate10randomNumbers = () => {
    let randomArray = [];
    for (let i = 0; i < 10; i++) {
      randomArray.push(Math.floor(Math.random() * 4))
    }
    return randomArray;
  }

  const handlePlayerResponse = colorNum => {
    playerResponse.push(colorNum);
    playerClickCounter++
    setPlayerClickDisplay(playerClickDisplay + 1);
    if (roundCount === playerClickCounter) {
      validateAnswer();
    }
  }

  return (
    <section>
      <ScoreBoard roundCount={roundCount} startGame={startGame} disableScoreBoard={disableScoreBoard} playerClickDisplay={playerClickDisplay} randomNumArray={randomNumArray} onViewRulesClick={onViewRulesClick} rulesVisible={rulesVisible}/>
      <div className={ rulesVisible ? 'rules-outer-container' : 'hidden' }>
        <div className='rules-inner-container'>
          <p className='rules-text'>
            It's you against Mr. Simon. When the game starts, Simon will create a color
            pattern that grows with each round. Your challenge is to match Simon's pattern
            by selecting the glowing lights in the correct order. Finish round 10 to win!
          </p>
        </div>
      </div>
      <div className={ gameBoardVisible ? 'gameboard-outer-container' : 'hidden' }>
        <div className='center-circle'>
          <p className='turn-text-1'>{myOrYours}</p>
          <p className='turn-text-2'>{turnText}</p>
        </div>
        <div className={ disableButtons ? 'top-section disabled' : 'top-section' }>
          <button aria-label='red button' onClick={() => handlePlayerResponse(0)} className={ redActive ? 'red-box red-active' : 'red-box' }></button>
          <button aria-label='yellow button' onClick={() => handlePlayerResponse(1)} className={ yellowActive ? 'yellow-box yellow-active' : 'yellow-box' }></button>
        </div>
        <div className={ disableButtons ? 'bottom-section disabled' : 'bottom-section' }>
          <button aria-label='green button' onClick={() => handlePlayerResponse(2)} className={ greenActive ? 'green-box green-active' : 'green-box' }></button>
          <button aria-label='blue button' onClick={() => handlePlayerResponse(3)} className={ blueActive ? 'blue-box blue-active' : 'blue-box' }></button>
        </div>
      </div>
    </section>
  );

}
