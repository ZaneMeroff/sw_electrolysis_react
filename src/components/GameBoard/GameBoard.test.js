import React from 'react';
import { GameBoard } from './GameBoard';
import { shallow } from 'enzyme';

describe('GameBoard', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<GameBoard />);
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

});
