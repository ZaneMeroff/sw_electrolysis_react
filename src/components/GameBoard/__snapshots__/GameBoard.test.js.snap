// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard should match the snapshot 1`] = `
<section>
  <ScoreBoard
    disableScoreBoard={false}
    onViewRulesClick={[Function]}
    playerClickDisplay={0}
    randomNumArray={Array []}
    roundCount={0}
    rulesVisible={false}
    startGame={[Function]}
  />
  <div
    className="hidden"
  >
    <div
      className="rules-inner-container"
    >
      <p
        className="rules-text"
      >
        It's you against Mr. Simon. When the game starts, Simon will create a color pattern that grows with each round. Your challenge is to match Simon's pattern by selecting the glowing lights in the correct order. Finish round 10 to win!
      </p>
    </div>
  </div>
  <div
    className="gameboard-outer-container"
  >
    <div
      className="center-circle"
    >
      <p
        className="turn-text-1"
      />
      <p
        className="turn-text-2"
      />
    </div>
    <div
      className="top-section disabled"
    >
      <button
        aria-label="red button"
        className="red-box"
        onClick={[Function]}
      />
      <button
        aria-label="yellow button"
        className="yellow-box"
        onClick={[Function]}
      />
    </div>
    <div
      className="bottom-section disabled"
    >
      <button
        aria-label="green button"
        className="green-box"
        onClick={[Function]}
      />
      <button
        aria-label="blue button"
        className="blue-box"
        onClick={[Function]}
      />
    </div>
  </div>
</section>
`;
